hedera-node/src/main/java/com/hedera/services/files/SpecialFileSystem.java

```diff
@@ -0,0 +1,229 @@
+			log.error(String.format("Error when reading fileID %s from local filesystem %s", fileID, e));
+			return new byte[0];
+		}
+	}
+
+	public synchronized void put(FileID fileID, byte[] content) {
+		try {
+			byte[] hash = MessageDigest.getInstance("SHA-384").digest(content);
+			log.info("New file size {} with hash {}", content.length, hex(hash));
+			fileMap.put(fileID, hash);
+			FileUtils.writeByteArrayToFile(
+					new File(SPECIAL_FILESYSTEM_DIR + fileSystemLocation + File.separator + fileIDtoDotString(fileID)),
+					content);
+			invalidateHash();
+		} catch (IOException e) {
+			log.error("Error when writing fileID {} to local filesystem", fileID, e);
+		} catch (NoSuchAlgorithmException e) {
+			log.error("Error when calculating hash of file {}", fileID);
+		}
+	}
+
+	public boolean isSpeicalFileID(FileID fileID) {
+		return fileMap.containsKey(fileID);
+	}
+
+	@Override
+	public void deserialize(SerializableDataInputStream in, int version) throws IOException {
+		int mapSize = in.readInt();
+		for (int i = 0; i < mapSize; i++) {
+			long realmID = in.readLong();
+			long shardID = in.readLong();
+			long fileNum = in.readLong();
+			FileID fileID = FileID.newBuilder()
+					.setFileNum(fileNum)
+					.setRealmNum(realmID)
+					.setShardNum(shardID).build();
+			byte[] hash = in.readByteArray(HASH_BYTES);
+			log.info("Read file {} with hash {}", fileID, hex(hash));
+			fileMap.put(fileID, hash);
+		}
+		invalidateHash();
+	}
+
+	@Override
+	public void serialize(SerializableDataOutputStream out) throws IOException {
+		out.writeInt(fileMap.size());
+		fileMap.keySet()
+				.stream()
+				.sorted(HederaLedger.FILE_ID_COMPARATOR)
+				.forEach(fileID -> {
+					try {
+						out.writeLong(fileID.getShardNum());
+						out.writeLong(fileID.getRealmNum());
+						out.writeLong(fileID.getFileNum());
+						out.writeByteArray(fileMap.get(fileID));
+					} catch (IOException e) {
+						log.error("Error when serialize {}", fileID);
+					}
+				});
+	}
+
+	@Override
+	public long getClassId() {
+		return RUNTIME_CONSTRUCTABLE_ID;
+	}
+
+	@Override
+	public int getVersion() {
+		return MERKLE_VERSION;
+	}
+
+	@Override
+	public SpecialFileSystem copy() {
+		SpecialFileSystem newCopy = new SpecialFileSystem(fileMap, this.fileSystemLocation);
+		return newCopy;
+	}
+
+	@Override
+	public void delete() {
+
+	}
+
+	public void verifyHash() {
+		for (FileID fileID : fileMap.keySet()) {
+			//Verify if the file system has the same hash loaded form state
+			byte[] hash = fileMap.get(fileID);
+			byte[] fileSystemHash = getFileHash(fileID);
+			if (!Arrays.equals(hash, fileSystemHash)) {
+				log.error("Error: File hash from state does not match file system, from state: {}", hex(hash));
+				log.error("Error: File hash from state does not match file system, from file : {}",
+						hex(fileSystemHash));
+			}
+		}
+	}
+
+	private boolean areEqualKeyValues(Map<FileID, byte[]> first, Map<FileID, byte[]> second) {
+		for (FileID key : first.keySet()) {
```

Comment:
Need to check that `first` and `second` have the same `FileID` keys. Addressed in #658.